runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp

```diff
@@ -103,16 +105,367 @@ void VMgenerateCatchBlockBBStartPrologue(TR::Node *node, TR::Instruction *fenceI
+
+   TR::Register *fieldClassReg = NULL;
+   bool isSideEffectReg = false;
+   // Load fieldClass
+   if (isInstanceField)
+      {
+      fieldClassReg = cg->allocateRegister();
+      generateLoadJ9Class(node, fieldClassReg, sideEffectRegister, cg);
+      }
+   else if (!(node->getSymbolReference()->isUnresolved()))
+      {
+      fieldClassReg = cg->allocateRegister();
+      if (cg->needClassAndMethodPointerRelocations())
+         {
+         // If this is an AOT compile, we generate instructions to load the fieldClass directly from the snippet because the fieldClass will be invalid
+         // if we load using the dataSnippet's helper query at compile time.
+         TR::MemoryReference *fieldClassMemRef = new (cg->trHeapMemory()) TR::MemoryReference(dataSnippetRegister, offsetof(J9JITWatchedStaticFieldData, fieldClass), cg);
+         generateTrg1MemInstruction(cg, TR::InstOpCode::ldrimmx, node, fieldClassReg, fieldClassMemRef);
+         }
+      else
+         {
+         // For non-AOT compiles we don't need to use sideEffectRegister here as the class information is available to us at compile time.
+         J9Class * fieldClass = static_cast<TR::J9WatchedStaticFieldSnippet *>(dataSnippet)->getFieldClass();
+         loadAddressConstant(cg, node, reinterpret_cast<intptr_t>(fieldClass), fieldClassReg);
+         }
+      }
+   else
+      {
+      // Unresolved
+      if (isWrite)
+         {
+         fieldClassReg = cg->allocateRegister();
+         generateTrg1MemInstruction(cg, TR::InstOpCode::ldrimmx, node, fieldClassReg,
+               new (cg->trHeapMemory()) TR::MemoryReference(sideEffectRegister, fej9->getOffsetOfClassFromJavaLangClassField(), cg));
+         }
+      else
+         {
+         isSideEffectReg = true;
+         fieldClassReg = sideEffectRegister;
+         }
+      }
+
+   TR::MemoryReference *classFlagsMemRef = new (cg->trHeapMemory()) TR::MemoryReference(fieldClassReg, static_cast<int32_t>(fej9->getOffsetOfClassFlags()), cg);
+
+   generateTrg1MemInstruction(cg, TR::InstOpCode::ldrimmw, node, scratchReg, classFlagsMemRef);
+   static_assert(J9ClassHasWatchedFields == 0x100, "We assume that J9ClassHasWatchedFields is 0x100");
+   generateTestImmInstruction(cg, node, scratchReg, 0x600); // 0x600 is immr:imms for 0x100
+   generateConditionalBranchInstruction(cg, TR::InstOpCode::b_cond, node, fieldReportLabel, TR::CC_NE);
+
+	generateReportOOL->swapInstructionListsWithCompilation();
+
+   generateLabelInstruction(cg, TR::InstOpCode::label, node, fieldReportLabel);
+   generateReportFieldAccessOutlinedInstructions(node, endLabel, dataSnippetRegister, isWrite, cg, sideEffectRegister, valueReg);
+
+   generateReportOOL->swapInstructionListsWithCompilation();
+
+   generateLabelInstruction(cg, TR::InstOpCode::label, node, endLabel);
+
+   cg->stopUsingRegister(scratchReg);
+   if (!isSideEffectReg)
+      cg->stopUsingRegister(fieldClassReg);
+
+   cg->machine()->setLinkRegisterKilled(true);
```

Comment:
I prefer this `setLinkRegisterKilled(true)` to be in `generateReportFieldAccessOutlinedInstructions()`, after generaing the `bl` instruction.

Reply:
Moved `setLinkRegisterKilled(true)` into `generateReportFieldAccessOutlinedInstructions()`.